panel/io/document.py

```diff
@@ -126,19 +128,20 @@ def _cleanup_doc(doc, destroy=True):
     # Destroy document
     doc.destroy(None)
 
-async def _run_write_futures(futures):
+async def _run_write_futures():
     """
     Ensure that all write_message calls are awaited and handled.
     """
     from tornado.websocket import WebSocketClosedError
     async with WRITE_LOCK:
-        for future in futures:
+        for future in _pending_writes[state._current_thread]:
```

Comment:
I don't think this is threadsafe in two ways, though I'm just looking at the code in isolation, and there could be other safeguards in other places. 

1. `_pending_writes[state._current_thread]` can add or remove items from the list during the run.
2. The `clear` below could clear a different list if `state._current_thread` has changed in another thread, as `asyncio.Lock()` is not threadsafe [[ref]](https://docs.python.org/3/library/asyncio-sync.html). A new item could also be added to the existing list, which is cleared with the rest (I'm not entirely sure if this is possible).

Reply:
>The clear below could clear a different list if state._current_thread has changed in another thread

Could you explain this? I'm not sure I'm following.

Reply:
Specifically I'm confused since `state._current_thread` cannot change (it's poorly named since it's not actually the current thread but the thread that the `curdoc` expects events to be dispatched from).